packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts

```diff
@@ -0,0 +1,337 @@
+		const states: Array<ProcessingState> = [];
+
+		for ( const state of this._processInProgress.values() ) {
+			if ( state.element.root.rootName == '$graveyard' ) {
+				states.push( state );
+			}
+		}
+
+		return states;
+	}
+
+	/**
+	 * Close image editor.
+	 */
+	private _handleImageEditorOnClose() {
+		if ( !this._wrapper ) {
+			return;
+		}
+
+		this._wrapper.remove();
+		this._wrapper = null;
+
+		this.editor.editing.view.focus();
+
+		this.refresh();
+	}
+
+	/**
+	 * Save edited image. In case server respond with "success" replace with edited image,
+	 * otherwise show notification error.
+	 */
+	private _handleImageEditorOnSave( state: ProcessingState, asset: CKBoxRawAssetDefinition ) {
+		const t = this.editor.locale.t;
+		const notification = this.editor.plugins.get( Notification );
+		const pendingActions = this.editor.plugins.get( PendingActions );
+		const action = pendingActions.add( t( 'Processing the edited image.' ) );
+
+		this._processInProgress.set( state.ckboxImageId, state );
+		this._showImageProcessingIndicator( asset );
+
+		this._waitForAssetProcessed( asset.data.id, state.controller.signal )
+			.then(
+				asset => {
+					this._replaceImage( state.element, asset );
+				},
+				() => {
+					if ( !state.controller.signal.aborted ) {
+						notification.showWarning( t( 'Server failed image processing or didn\'t respond' ), {
+							title: t( 'Server failed image processing or didn\'t respond' ),
+							namespace: 'ckbox'
+						} );
+					}
+
+					// Remove processing indicator. It was added only to ViewElement.
+					this.editor.editing.reconvertItem( state.element );
+				}
+			).finally( () => {
+				this._processInProgress.delete( state.ckboxImageId );
+				pendingActions.remove( action );
+			} );
+
+		this.refresh();
+	}
+
+	/**
+	 * Get asset's status on server. If server responds with "success" status then
+	 * image is already proceeded and ready for saving.
+	 */
+	private async _getAssetStatusFromServer( id: string, signal: AbortSignal ): Promise<CKBoxRawAssetDefinition> {
+		const url = new URL( 'assets/' + id, this.editor.config.get( 'ckbox.serviceOrigin' )! );
+		const ckboxEditing = this.editor.plugins.get( CKBoxEditing );
+		const response: CKBoxRawAssetDataDefinition = await sendHttpRequest( {
+			url,
+			signal,
+			authorization: ckboxEditing.getToken().value
+		} );
+		const status = response.metadata!.metadataProcessingStatus;
+
+		if ( !status || status == 'queued' ) {
+			throw new Error( 'Image has not been processed yet.' );
+		}
+
+		return { data: { ...response } };
+	}
+
+	private async _waitForAssetProcessed( id: string, signal: AbortSignal ): Promise<CKBoxRawAssetDefinition> {
+		const result = await retry(
+			() => this._getAssetStatusFromServer( id, signal ),
+			{
+				signal,
+				maxAttempts: 5
+			}
+		);
+
+		if ( result.data.metadata!.metadataProcessingStatus != 'success' ) {
+			throw new Error( 'Image processing failed.' );
```

Comment:
We should use the `CKEditorError` to be able to distinguish runtime exceptions from known use cases.

Reply:
I mean the case of `Uncaught TypeError: Cannot read properties of undefined (reading 'foo')` should not be displayed as a notification. It should be logged to the console as an error for a developer to fix. It's a code bug, not a known use case that the user should be informed of and would be able to solve.